Data/Graph/Inductive/Query/Cycles.hs

```diff
@@ -0,0 +1,178 @@
+      split = zip ns tns'
+      tns' = tail $ tails ns
+      allTwoCycle (n,rs) = null $ rs \\ twoCycle g n
+
+-- End of copied functions.
+
+
+
+-- | Contains the necessary data structures used by 'cycles'.
+data CyclesInState g a b
+  = CyclesInState
+      { cisCycles :: [[Node]]
+        -- ^ The cycles found so far, in topological order.
+      , cisBlocked :: M.IntMap Bool
+        -- ^ The nodes which are currently blocked.
+      , cisBlockMap :: M.IntMap [Node]
+        -- ^ The B set.
+      , cisStack :: [Node]
+        -- ^ The node stack.
+      , cisS :: Maybe Node
+        -- ^ The current S value.
+      , cisCurrentComp :: Maybe (g a b)
+        -- ^ The component currently being processed.
+      , cisComponents :: [g a b]
+        -- ^ The components of the input graph.
+      , cisGraph :: g a b
+        -- ^ The input graph.
+      }
+  deriving (Show, Read, Eq)
+
+-- | Finds all cycles in a given graph. The returned lists contains the nodes
+-- appearing in the cycles, in successor order (although it is undefined which
+-- node appears first).
+--
+-- Implemented using Johnson's algorithm. See Donald B. Johnson: Finding All the
+-- Elementary Circuits of a Directed Graph. SIAM Journal on Computing. Volumne
+-- 4, Nr. 1 (1975), pp. 77-84.
+cycles :: (Graph g) => g a b -> [[LNode a]]
+cycles g = map (addLabels g) (cycles' g)
+
+-- | Same as 'cycles' but for unlabeled graphs.
+cycles' :: (Graph g) => g a b -> [[Node]]
+cycles' g =
+  cisCycles $
+  foldr cyclesFor (mkInitCyclesInState g) (nodes g)
+
+-- | Find all cycles in the given graph (using 'cycles'), excluding those that
+-- are also cliques.
+uniqueCycles :: (Graph g) => g a b -> [[LNode a]]
+uniqueCycles g = map (addLabels g) (uniqueCycles' g)
+
+-- | Same as 'uniqueCycles' but for unlabeled graphs.
+uniqueCycles' :: (Graph g) => g a b -> [[Node]]
+uniqueCycles' g = filter (not . isRegular g) (cycles' g)
+
+cyclesFor :: (Graph g) => Node -> CyclesInState g a b -> CyclesInState g a b
+cyclesFor n st0 =
+  let n_comp = head $
+               filter (\c -> n `gelem` c) $
+               cisComponents st0
+  in if noNodes n_comp > 1
+     then let st1 = st0 { cisS = Just n
+                        , cisCurrentComp = Just n_comp
+                        }
+              st2 = fst $ cCircuits n st1
+              g = cisGraph st2
+              new_g = delNode n g
+              new_comps = strongComponentsOf new_g
+              st3 = st2 { cisGraph = new_g
+                        , cisComponents = new_comps
+                        }
+          in st3
+     else st0 -- Skip to next node
+
+cCircuits :: (Graph g) => Node -> CyclesInState g a b ->
+             (CyclesInState g a b, Bool)
+cCircuits n st0 =
+  let st1 = st0 { cisBlocked = M.insert n True (cisBlocked st0)
+                , cisStack = (n:cisStack st0)
+                }
+      c = fromJust $ cisCurrentComp st1
+      n_suc = suc c n
+      (st2, f) =
+        foldr ( \m (st, f') ->
```

Comment:
This is big enough that it should be split into a separate function.

Reply:
Now broken into several, smaller functions.